kymatio/scattering2d/backend/tensorflow_backend.py

```diff
@@ -0,0 +1,165 @@
+            tensor such that its fourier transform is the Fourier
+            transform of a subsampled version of x, i.e. in
+            FFT^{-1}(res)[u1, u2] = FFT^{-1}(x)[u1 * (2**k), u2 * (2**k)]
+    """
+    def __call__(self, x, k):
+        y = tf.reshape(x,(-1, k, x.shape[1] // k, k, x.shape[2] // k))
+        
+        out = tf.reduce_mean(y, axis=(1, 3))
+        print(tf.shape(out))
+        return out
+
+
+class Modulus(object):
+    """
+        This class implements a modulus transform for complex numbers.
+        Usage
+        -----
+        modulus = Modulus()
+        x_mod = modulus(x)
+        Parameters
+        ---------
+        x: input complex tensor.
+        Returns
+        -------
+        output: a real tensor equal to the modulus of x.
+    """
+    def __call__(self, x):
+        norm = tf.abs(x)
+        return tf.cast(norm, tf.complex64)
+
+
+def fft(x, direction='C2C', inverse=False):
+    """
+        Interface with torch FFT routines for 2D signals.
+        Example
+        -------
+        x = torch.randn(128, 32, 32, 2)
+        x_fft = fft(x, inverse=True)
+        Parameters
+        ----------
+        input : tensor
+            complex input for the FFT
+        direction : string
+            'C2R' for complex to real, 'C2C' for complex to complex
+        inverse : bool
+            True for computing the inverse FFT.
+            NB : if direction is equal to 'C2R', then an error is raised.
+    """
+    if direction == 'C2R':
+        if not inverse:
+            raise RuntimeError('C2R mode can only be done with an inverse FFT.')
+
+    if direction == 'C2R':
+        output = tf.math.real(tf.signal.ifft2d(x, name='irfft2d'))*tf.cast(x.shape[-1]*x.shape[-2],tf.float32)
+    elif direction == 'C2C':
+        if inverse:
+            output = tf.signal.ifft2d(x, name='ifft2d')*tf.cast(x.shape[-1]*x.shape[-2], tf.complex64)
+        else:
+            output = tf.signal.fft2d(x, name='fft2d')
+
+    return output
+
+
+def cdgmm(A, B, inplace=False):
+    """
+        Complex pointwise multiplication between (batched) tensor A and tensor B.
+        Parameters
+        ----------
+        A : tensor
+            A is a complex tensor of size (B, C, M, N, 2)
+        B : tensor
+            B is a complex tensor of size (M, N) or real tensor of (M, N)
+        inplace : boolean, optional
+            if set to True, all the operations are performed inplace
+        Returns
+        -------
+        C : tensor
+            output tensor of size (B, C, M, N, 2) such that:
+            C[b, c, m, n, :] = A[b, c, m, n, :] * B[m, n, :]
+    """
+    if B.ndim != 2:
+        raise RuntimeError('The dimension of the second input must be 2.')
+    print(tf.shape(A))
+    return A * B
+
+def empty_like(x, shape):
+    return tf.zeros(shape, dtype=x.dtype)
+
+def squeeze(x, y, index):
+    y_list = tf.unstack(y, axis = -3)
+    y_list[index] = tf.squeeze(x, axis = -3)
+    return tf.stack(y_list, axis = -3)
```

Comment:
this seems like SUCH a large amount of memory being moved around if `y` was large...

Reply:
I don't know how tf works.

Reply:
I think none of us if a tf user, would it be OK to wait for tf user to improve our code?(maybe @janden wants to dedicate some times to optimize this..)

Reply:
Might be a good idea. 

I'll try to spend some time tomorrow to see whats the best way of doing this stuff. (right now I should sleep, 2am almost lol )

Reply:
Yeah I'm fine leaving this somewhat unoptimized for now. Important thing is to have something that works and has an interface that we believe will be stable for the next few versions.

Reply:
As discussed in #468 , we will be bringing back concatenate.